homeassistant/components/notify/telegram.py

```diff
@@ -81,110 +105,254 @@ def load_data(url=None, file=None, username=None, password=None):
-            return self.send_document(data.get(ATTR_DOCUMENT))
-        elif data is not None and ATTR_KEYBOARD in data:
-            keys = data.get(ATTR_KEYBOARD)
-            keys = keys if isinstance(keys, list) else [keys]
-            return self.send_keyboard(message, keys)
-
-        if title:
-            text = '{} {}'.format(title, message)
+        self._default_user = None
+        self._users = {}
+        for i, (dev_id, user_id) in enumerate(user_id_array.items()):
+            if i == 0:
+                self._default_user = user_id
+            self._users[user_id] = dev_id
+        self._parsers = {PARSER_HTML: ParseMode.HTML,
+                         PARSER_MD: ParseMode.MARKDOWN}
+        self._parse_mode = self._parsers.get(parser)
+        # last_msg_cache:
+        self._last_message_id = {user: None for user in self._users.keys()}
+        self.bot = Bot(token=self._api_key)
+
+    def _get_msg_ids(self, msg_data, chat_id):
+        """Get the message id to edit.
+
+        This can be one of (message_id, inline_message_id) from a msg dict,
+        returning a tuple.
+        **You can use 'last' as message_id** to edit
+        the last sended message in the chat_id.
+        """
+        message_id = inline_message_id = None
+        if 'message_id' in msg_data:
+            message_id = msg_data['message_id']
+            if (isinstance(message_id, str) and (message_id == 'last') and
+                    (self._last_message_id[chat_id] is not None)):
+                message_id = self._last_message_id[chat_id]
         else:
-            text = message
-
-        parse_mode = telegram.parsemode.ParseMode.MARKDOWN
-
-        # send message
-        try:
-            self.bot.sendMessage(chat_id=self._chat_id,
-                                 text=text,
-                                 parse_mode=parse_mode)
-        except telegram.error.TelegramError:
-            _LOGGER.exception("Error sending message")
-
-    def send_keyboard(self, message, keys):
-        """Display keyboard."""
-        import telegram
-
-        keyboard = telegram.ReplyKeyboardMarkup([
-            [key.strip() for key in row.split(",")] for row in keys])
+            inline_message_id = msg_data['inline_message_id']
+        return message_id, inline_message_id
+
+    def _get_target_chat_ids(self, target):
+        """Validate chat_id targets or return default target (fist defined).
+
+        :param target: optional list of strings (['12234'])
+        :return list of chat_id targets (integers)
+        """
+        if target is not None:
+            if isinstance(target, int):
+                if target in self._users:
+                    return [target]
+                _LOGGER.warning('BAD TARGET "%s", using default: %s',
+                                target, self._default_user)
+            else:
+                try:
+                    chat_ids = [int(t) for t in target if int(t) in self._users]
```

Comment:
I'm testing the requested changes, I hope tomorrow I'll have the time to finish the PR and the  associated doc.

I have limited the configuration of a single platform ('webhooks' or 'polling') within the `telegram_bot` component, linking the notification services to the name 'telegram_bot / service_name', according to the previous comment.

The notify.telegram component now depends on `telegram_bot` and does not require the python-telegram module. Now it is only an optional "fast link" to the new notification services of the telegram_bot component, but allows backward compatibility for all service calls to `notify.telegram` with the old syntax.
Basically, it allows you to generate a customised shortcut to send notifications to a particular chat_id.
It would look like this:
```python
"""
Telegram platform for notify component.

For more details about this platform, please refer to the documentation at
https://home-assistant.io/components/notify.telegram/
"""
import logging
import voluptuous as vol

import homeassistant.helpers.config_validation as cv
from homeassistant.components.notify import (
    ATTR_MESSAGE, ATTR_TITLE, ATTR_DATA, ATTR_TARGET,
    PLATFORM_SCHEMA, BaseNotificationService)
from homeassistant.const import ATTR_LOCATION

_LOGGER = logging.getLogger(__name__)

DOMAIN = 'telegram_bot'
DEPENDENCIES = [DOMAIN]

ATTR_KEYBOARD = 'keyboard'
ATTR_INLINE_KEYBOARD = 'inline_keyboard'
ATTR_PHOTO = 'photo'
ATTR_DOCUMENT = 'document'

CONF_CHAT_ID = 'chat_id'

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Required(CONF_CHAT_ID): cv.positive_int,
})


def get_service(hass, config, discovery_info=None):
    """Get the Telegram notification service."""
    chat_id = config.get(CONF_CHAT_ID)
    return TelegramNotificationService(hass, chat_id)


class TelegramNotificationService(BaseNotificationService):
    """Implement the notification service for Telegram."""

    def __init__(self, hass, chat_id):
        """Initialize the service."""
        self._chat_id = chat_id
        self.hass = hass

    def send_message(self, message="", **kwargs):
        """Send a message to a user."""

        service_data = dict(target=kwargs.get(ATTR_TARGET, self._chat_id))
        if ATTR_TITLE in kwargs:
            service_data.update({ATTR_TITLE: kwargs.get(ATTR_TITLE)})
        if message:
            service_data.update({ATTR_MESSAGE: message})
        data = kwargs.get(ATTR_DATA)

        # Get keyboard info
        if data is not None and ATTR_KEYBOARD in data:
            keys = data.get(ATTR_KEYBOARD)
            keys = keys if isinstance(keys, list) else [keys]
            service_data.update(keyboard=keys)
        elif data is not None and ATTR_INLINE_KEYBOARD in data:
            keys = data.get(ATTR_INLINE_KEYBOARD)
            keys = keys if isinstance(keys, list) else [keys]
            service_data.update(inline_keyboard=keys)

        # Send a photo, a document or a location
        if data is not None and ATTR_PHOTO in data:
            photos = data.get(ATTR_PHOTO, None)
            photos = photos if isinstance(photos, list) else [photos]
            for photo_data in photos:
                service_data.update(photo_data)
                self.hass.services.call(
                    DOMAIN, 'send_photo', service_data=service_data)
            return
        elif data is not None and ATTR_LOCATION in data:
            service_data.update(data.get(ATTR_LOCATION))
            return self.hass.services.call(
                DOMAIN, 'send_location', service_data=service_data)
        elif data is not None and ATTR_DOCUMENT in data:
            service_data.update(data.get(ATTR_DOCUMENT))
            return self.hass.services.call(
                DOMAIN, 'send_document', service_data=service_data)

        # Send message
        _LOGGER.debug('TELEGRAM NOTIFIER calling %s.send_message with %s',
                      DOMAIN, service_data)
        return self.hass.services.call(
            DOMAIN, 'send_message', service_data=service_data)
```